staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -348,6 +359,19 @@ func (r *crdHandler) getServingInfoFor(crd *apiextensions.CustomResourceDefiniti
 	return ret
 }
 
+func (c *crdHandler) updateCustomResourceDefinition(oldObj, newObj interface{}) {
+	oldCRD := oldObj.(*apiextensions.CustomResourceDefinition)
+	newCRD := newObj.(*apiextensions.CustomResourceDefinition)
+	glog.V(4).Infof("Updating customresourcedefinition %s", oldCRD.Name)
+
+	c.customStorageLock.Lock()
+	defer c.customStorageLock.Unlock()
+
+	storageMap := c.customStorage.Load().(crdStorageMap)
+	delete(storageMap, oldCRD.UID)
+	_ = c.getServingInfoFor(newCRD)
```

Comment:
I fear we have a data race now: `getServingInfoFor` reads a value from the map, here we delete it. On the other hand, we don't want to move the `Lock()` up in `getServingInfoFor`. I think we need a RWMutex, or we make a copy of the map here, change it and write the map back.